File: Greedy/MinOpsToReduceNumToZero/min.py
class Solution:
    def doDP(self, n: int, dp: list[int]) -> int:
        if bin(n).count("1") == 1:
            return 0

        if dp[n] != -1:
            return dp[n]

        if n % 2 == 0:
            dp[n] = self.doDP(n // 2, dp)
        else:
            dp[n] = 1 + min(
                self.doDP(n - 1, dp),
                self.doDP(n + 1, dp)
            )

        return dp[n]

    def minOperations(self, n: int) -> int:
        dp = [-1] * (n + 2)
        return self.doDP(n, dp) + 1

File: Greedy/MinOpsToReduceNumToZero/min.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int doDP(int n,vector<int>& dp){
        if(__builtin_popcount(n)==1)return 0;
        else if(dp[n]!=-1)return dp[n];
        else if(n%2==0){
            return dp[n] = doDP(n/2,dp);
        }
        return dp[n] = 1+min(doDP(n-1,dp),doDP(n+1,dp));
    }
    int minOperations(int n) {
        vector<int>dp(n+2,-1);
        return doDP(n,dp)+1;
    }
};